### PR TITLE
[Enhancement] Use the native scanner to read hudi MOR table without log files for read optimized queries

### PR DESCRIPTION
**Pull request by miomiocat**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

1. use the native reader to read hudi MOR table without log files for read optimized queries instead of the jni reader
2. jni reader will only be used in this case: hudi MOR table, snapshot queries and log files exist
3. some code refinement and bugfix

### CODE REVIEW

**Comment by miomiocat** [comment]
run starrocks_fe_unittest

**Review by dirtysalt — DISMISSED** [review]

**Comment by miomiocat** [comment]
run starrocks_fe_unittest

**Comment by miomiocat** [comment]
run starrocks_fe_unittest

**Comment by miomiocat** [comment]
run starrocks_fe_unittest

**Comment by miomiocat** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 12 (75.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|1|2|50.00%|[337]|
|:large_blue_circle:|com/starrocks/catalog/HudiTable.java|6|8|75.00%|[127, 546]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaCache.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/external/hive/HdfsFileDesc.java|1|1|100.00%|[]|

**Review by caneGuy — DISMISSED** [review]

File: be/src/connector/hive_connector.cpp [hunk 1]
```diff
@@ -259,7 +259,25 @@ Status HiveDataSource::_init_scanner(RuntimeState* state) {
...
+
+    bool use_hudi_jni_reader = false;
+    if (dynamic_cast<const HudiTableDescriptor*>(_tuple_desc->table_desc())) {
+        const auto* hudi_table = dynamic_cast<const HudiTableDescriptor*>(_hive_table);
+        bool is_mor_table = hudi_table->is_mor_table();
+        bool snapshot_mode =
+                hudi_table->get_input_format() == "org.apache.hudi.hadoop.realtime.HoodieParquetRealtimeInputFormat";
+        bool read_optimized_mode = hudi_table->get_input_format() == "org.apache.hudi.hadoop.HoodieParquetInputFormat";
+        // logs only table
+        if (is_mor_table && read_optimized_mode && scan_range.file_length == -1 && scan_range.relative_path.empty()) {
```
Comment (DorianZheng) [inline be/src/connector/hive_connector.cpp:271]: Can we filter this kind of scan range in fe

File: be/src/connector/hive_connector.cpp [hunk 2]
```diff
@@ -259,7 +259,25 @@ Status HiveDataSource::_init_scanner(RuntimeState* state) {
 
     HdfsScanner* scanner = nullptr;
     auto format = scan_range.file_format;
-    if (dynamic_cast<const HudiTableDescriptor*>(_tuple_desc->table_desc()) && scan_range.hudi_mor_table) {
+
+    bool use_hudi_jni_reader = false;
```
Comment (DorianZheng) [inline be/src/connector/hive_connector.cpp:263]: Is it better to move this logic from be to fe to simplify be

**Comment by DorianZheng** [comment]
@miomiocat I this it's better to move the knowledge of snapshot/ro from be to fe, since fe itself have to understand it anyway, but be can know nothing about it which simplify be code.

File: be/src/connector/hive_connector.cpp (hunk 2 above)
Comment (miomiocat) [reply be/src/connector/hive_connector.cpp:263]: done

File: be/src/connector/hive_connector.cpp (hunk 1 above)
Comment (miomiocat) [reply be/src/connector/hive_connector.cpp]: done

**Comment by miomiocat** [comment]
> @miomiocat I this it's better to move the knowledge of snapshot/ro from be to fe, since fe itself have to understand it anyway, but be can know nothing about it which simplify be code.

done, PTAL

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 22 (40.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/RemoteScanRangeLocations.java|0|9|00.00%|[134, 180, 181, 182, 183, 192, 193, 195, 196]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|1|2|50.00%|[337]|
|:large_blue_circle:|com/starrocks/catalog/HudiTable.java|6|9|66.67%|[127, 135, 550]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaCache.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/external/hive/HdfsFileDesc.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/external/RemoteScanRangeLocations.java [hunk 3]
```diff
@@ -175,14 +177,23 @@ public void setupScanRangeLocations(DescriptorTable descTbl, Table table,
                 }
             }
         } else if (table instanceof HudiTable) {
+            HudiTable hudiTable = (HudiTable) table;
+            boolean readOptimized = hudiTable.getHudiInputFormat().equals("org.apache.hudi.hadoop.HoodieParquetInputFormat");
+            boolean snapshot = hudiTable.getHudiInputFormat().equals("org.apache.hudi.hadoop.HoodieParquetInputFormat");
```
[the remark this answers is not in the document]
Comment (miomiocat) [reply fe/fe-core/src/main/java/com/starrocks/external/RemoteScanRangeLocations.java]: typo, fixed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/RemoteScanRangeLocations.java|0|12|00.00%|[134, 180, 181, 182, 183, 184, 185, 186, 195, 196, 198, 199]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|1|2|50.00%|[337]|
|:large_blue_circle:|com/starrocks/catalog/HudiTable.java|6|9|66.67%|[134, 142, 557]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaCache.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/external/hive/HdfsFileDesc.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/RemoteScanRangeLocations.java|0|12|00.00%|[134, 180, 181, 182, 183, 184, 185, 186, 195, 196, 198, 199]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|1|2|50.00%|[337]|
|:large_blue_circle:|com/starrocks/catalog/HudiTable.java|6|9|66.67%|[134, 142, 557]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaCache.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/external/hive/HdfsFileDesc.java|1|1|100.00%|[]|

**Review by caneGuy — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by miomiocat** [comment]
run starrocks_admit_test

**Comment by miomiocat** [comment]
run starrocks_admit_test

**Pull request merged**